### PR TITLE
stdlib unicode string additions

### PR DESCRIPTION
**Pull request by dgud**
Here is a suggestion for a new unicode handling and a new string module in stdlib that operates on `unicode:chardata()`, thus handles both binaries and lists of characters.

The following conversion functions are added to the unicode module:
```erlang
%% Canonical decompose string to list of chars
-spec characters_to_nfd_list(chardata()) -> [char()].
1> unicode:characters_to_nfd_list("abc..åäö"). 
[97,98,99,46,46,97,778,97,776,111,776]
```
```erlang
-spec characters_to_nfd_binary(chardata()) -> unicode_binary().
2> unicode:characters_to_nfd_binary("abc..åäö").
<<97,98,99,46,46,97,204,138,97,204,136,111,204,136>>
```
```erlang
%% Compability Canonical decompose string to list of chars.
-spec characters_to_nfkd_list(chardata()) -> [char()].
3> unicode:characters_to_nfkd_list("３２").
"32"
```
```erlang
-spec characters_to_nfkd_binary(chardata()) -> unicode_binary().

%% Canonical compose string to list of chars
-spec characters_to_nfc_list(chardata()) -> [char()].
-spec characters_to_nfc_binary(chardata()) -> unicode_binary().

%% Compability Canonical compose string to list of chars
-spec characters_to_nfkc_list(chardata()) -> [char()].
-spec characters_to_nfkc_binary(chardata()) -> unicode_binary().
```

The new string module works with unicode:chardata() as input as was decided on OTP board
meeting as response to [EEP-35](http://www.erlang.org/erlang-enhancement-proposals/eep-0035.html) a long time ago. Furthermore improves erlangs unicode handling by operating on grapheme clusters.

The new string module currently called `str`, operates on grapheme clusters, 
though it does not handle extended grapheme clusters, i.e. prepend class,
nor classic or "nfd" Hangul.
A pragmatic decision to get some performance when searching in binary input.

This implementation does not, currently, address language or other context specifics, 
which requires domain knowledge or hard coding rules in the code generator.

Most functions expect all input to be normalized to the same form.

The return value type are the same as the input value type, unless the result type
is explicit specified, e.g.
```erlang
5> str:split("ab..bc..cd", "..", leading).
["ab","bc..cd"]
6> str:split(<<"ab..bc..cd">>, "..", leading).
[<<"ab">>,<<"bc..cd">>]

7> str:join(str:split(<<"ab..bc..cd">>, "..", trailing), "..").
[<<"ab..bc">>,"..",<<"cd">>]
```
 str.erl API

```erlang
-type grapheme_cluster() :: char() | [char()].
-type direction() :: 'leading' | 'trailing'.

%% Check if string is the empty string
-spec is_empty(unicode:chardata()) -> boolean().
7> str:is_empty(["", <<>>]).
true

%% Count the number of grapheme clusters in chardata
-spec length(unicode:chardata()) -> non_neg_integer().
8> str:length("abc..åäö").                                                               
8
9> str:length(lists:flatten([97,98,99,46,46,[97,778],[97,776],[111,776]])).
8

%% Convert a string to a list of grapheme clusters
-spec to_graphemes(unicode:chardata()) -> [grapheme_cluster()].
10> str:to_graphemes("abc..åäö").   
"abc..åäö"
11> str:to_graphemes(unicode:characters_to_nfd_list("abc..åäö")).
[97,98,99,46,46,[97,778],[97,776],[111,776]]

%% Reverse grapheme clusters
-spec reverse(unicode:chardata()) -> [grapheme_cluster()].
12> str:reverse(unicode:characters_to_nfd_list("abc..åäö")).     
[[111,776],[97,776],[97,778],46,46,99,98,97]

%% Append a character or string to a strings.
%% Normally we just append with Str = [Str1|Str2] or [Str1,Str2]
%% But if we want to keep a certain type we can use concat which copies the data.
-spec concat(A::grapheme_cluster()|unicode:chardata(),B::unicode:chardata()) -> unicode:chardata().
14> str:concat(<<"abc.">>, <<".åäö"/utf8>>).
<<"abc..åäö"/utf8>>
16> str:concat("abc.", ".åäö").     
"abc..åäö"

%% Compare two strings return boolean, assumes that the input are
%% normalized to same form, see unicode:characters_to_nfX_xxx(..)
-spec equal(A::unicode:chardata(), B::unicode:chardata()) -> boolean().
17> str:equal("abc..åäö", <<"abc..åäö"/utf8>>). 
true
18> str:equal("abc..åäö", unicode:characters_to_nfd_list(<<"abc..åäö"/utf8>>)). 
false

%% Compare two strings return boolean, assumes that the input are
%% normalized to same form, see unicode:characters_to_nfX_xxx(..)
%% does casefold on the fly
-spec equal(A::unicode:chardata(), B::unicode:chardata(), IgnoreCase::boolean()) -> boolean().
19> str:equal("abc..åäö", <<"ABC..åÄö"/utf8>>, true).                           
true

%% Compare two strings return boolean
%% if specified does casefold and normalization on the fly
-spec equal(A, B, IgnoreCase, Norm) -> boolean() when
      A :: unicode:chardata(),
      B :: unicode:chardata(),
      IgnoreCase :: boolean(),
      Norm :: 'none' | 'nfc' | 'nfd' | 'nfkc' | 'nfkd'.
20> str:equal("abc..åäö", unicode:characters_to_nfd_list(<<"ABC..åÄö"/utf8>>), true, nfc).
true

%% Slice a string and return rest of string
%% Note: counts grapheme_clusters
-spec slice(unicode:chardata(), Start) -> Slice when
      Start :: non_neg_integer(),
      Slice :: unicode:chardata().
21> str:slice("abc..åäö", 4).                                                             
".åäö"
22> str:slice(<<"abc..åäö"/utf8>>, 4).
<<".åäö"/utf8>>

-spec slice(unicode:chardata(), Start, Length) -> Slice when
      Start :: non_neg_integer(),
      Length :: 'infinity' | non_neg_integer(),
      Slice :: unicode:chardata().
23> str:slice(<<"abc..åäö"/utf8>>, 2, 4).
<<"c..å"/utf8>>

%% Intersperse a list of strings with Sep
-spec join([Str::unicode:chardata()], Sep::unicode:chardata()) -> [unicode:chardata()].
24> str:join([<<"abc">>, <<"åäö"/utf8>>], "..").
[<<"abc">>,"..",<<"åäö"/utf8>>]

%% Pad a string to desired length
-spec pad(unicode:chardata(), Length::integer()) -> unicode:charlist().
-spec pad(unicode:chardata(), Length::integer(), Dir::direction()|'center') -> unicode:charlist().
-spec pad(Str, Length, Dir, Char) -> unicode:charlist() when
      Str ::unicode:chardata(),
      Length :: integer(),
      Dir :: direction()|'center',
      Char :: grapheme_cluster().
25> str:pad(<<"abc..åäö"/utf8>>, 10, center, $+). 
["+",<<"abc..åäö"/utf8>>,"+"]

%%  Strip characters from whitespace or Separator in Direction
-spec strip(Str::unicode:chardata()) -> unicode:chardata().
-spec strip(Str::unicode:chardata(),direction()|'both') -> unicode:chardata().
-spec strip(Str::unicode:chardata(),direction()|'both',[grapheme_cluster()]) ->
                   unicode:chardata().
31> str:strip("  foo  \n").                        
"foo"
32> str:strip("  foo  \n", leading).
"foo  \n"
36> str:strip(<<"  foo ++ ">>, both, [$+,$\s]).
<<"foo">>

%% Delete trailing newlines or \r\n
-spec chomp(Str::unicode:chardata()) -> unicode:chardata().
37> str:chomp("  foo  \n").                    
"  foo  "

%% Uppercase all chars in Str
-spec uppercase(Str::unicode:chardata()) -> unicode:chardata().
26> str:uppercase("Michał").
[77,73,67,72,65,321]

%% Lowercase all chars in Str
-spec lowercase(Str::unicode:chardata()) -> unicode:chardata().
27> str:lowercase(str:uppercase("michał")).
[109,105,99,104,97,322]

%% Make a titlecase of the first char in Str
-spec titlecase(Str::unicode:chardata()) -> unicode:chardata().
29> io:format("~ts~n",[str:titlecase("ǉusad")]).
ǈusad

%% Make a comparable string of the Str should be used for equality tests only
-spec casefold(Str::unicode:chardata()) -> unicode:chardata().
32> io:format("~ts~n",[str:casefold("ẞ SHARP S")]). 
ss sharp s

%% Return the remaining string with prefix removed or else nomatch
-spec prefix(Str::unicode:chardata(), Prefix::unicode:chardata()) -> nomatch | unicode:chardata().
57> str:prefix("prefix", "pre").                                                     
"fix"

%% split Haystack with the first occurrence of Needle, return list of splits
-spec split(Haystack::unicode:chardata(),
            Needle::unicode:chardata()) ->
                   [unicode:chardata()].

%% split Haystack with Needle, return list of splits
-spec split(Haystack::unicode:chardata(),
            Needle::unicode:chardata(),
            Where::direction()|'all') ->
                   [unicode:chardata()].
5> str:split("ab..bc..cd", "..", leading).
["ab","bc..cd"]
6> str:split(<<"ab..bc..cd">>, "..", trailing).
[<<"ab..bc">>,<<"cd">>]

%% Replace the first Needle in Haystack with Replacement
-spec replace(Haystack::unicode:chardata(),
              Needle::unicode:chardata(),
              Replacement::unicode:chardata()) -> [unicode:chardata()].
%% Replace Where Needle in Haystack with Replacement
-spec replace(Haystack::unicode:chardata(),
              Needle::unicode:chardata(),
              Replacement::unicode:chardata(),
              Where::direction()|'all') -> [unicode:chardata()].
46> str:replace("ab..cd..ef", "..", "*").       
["ab","*","cd..ef"]
47> str:replace(<<"ab..cd..ef">>, "..", "*", all).
[<<"ab">>,"*",<<"cd">>,"*",<<"ef">>]

%% Split Str into a list of chardata separated by one of the grapheme clusters in Seps
-spec tokens(Str::unicode:chardata(), Seps::[grapheme_cluster()]) -> [unicode:chardata()].
35> str:tokens(<<"ab..cd..ef">>, ".").  
[<<"ab">>,<<"cd">>,<<"ef">>]

%% find first Needle in Haystack return rest of string
-spec find(Haystack::unicode:chardata(), Needle::unicode:chardata()) -> unicode:chardata().
%% find Needle in Haystack (search in Dir direction) return rest of string
-spec find(Haystack::unicode:chardata(), Needle::unicode:chardata(), Dir::direction()) -> unicode:chardata().
49> str:find("ab..cd..ef", ".").  
"..cd..ef"
50> str:find(<<"ab..cd..ef">>, ".", trailing).
<<".ef">>

%% Fetch first codepoint and return rest in tail
-spec gc(Str::unicode:chardata()) -> maybe_improper_list(grapheme_cluster(),unicode:chardata()).
gc(CD) -> unicode_util:gc(CD).
53> str:gc([111,776,97,776,97,778,46,46,99,98,97]).
[[111,776],97,776,97,778,46,46,99,98,97]
56> str:gc(unicode:characters_to_binary([111,776,97,776,97,778,46,46,99,98,97])).
[[111,776]|<<97,204,136,97,204,138,46,46,99,98,97>>]

%% Fetch first grapheme cluster and return rest in tail
-spec cp(Str::unicode:chardata()) -> maybe_improper_list(char(),unicode:chardata()).
cp(CD) -> unicode_util:cp(CD).
52> str:cp([111,776,97,776,97,778,46,46,99,98,97]).
[111,776,97,776,97,778,46,46,99,98,97]
55> str:cp(unicode:characters_to_binary([111,776,97,776,97,778,46,46,99,98,97])).
[111|<<204,136,97,204,136,97,204,138,46,46,99,98,97>>]

```

### CODE REVIEW

**Comment by isvilen** [comment]
I think that it will be very useful to provide full access to information in Unicode Character Database (UnicodeData.txt, PropList.txt, SpecialCasing.txt, ... etc.), so that user applications can implement other aspects of Unicode standard (words and lines segmentation, bidirectional layout algorithm, language specific tailoring etc.) without to have to start from scratch.

I have a [proof of concept library](https://github.com/isvilen/unicodedata) that uses parse transform to generate code points information functions from Unicode Character Database files, for example the implementations of ```is_lowercase/1``` and  ```is_case_ignorable/1``` functions are:
``` erlang
-spec is_lowercase(char()) -> boolean().
is_lowercase(CP) ->
    ucd_is_category(CP, 'Ll') orelse ucd_has_property(CP, other_lowercase).

-spec is_case_ignorable(char()) -> boolean().
is_case_ignorable(CP) ->
    ucd_word_break(CP, [mid_letter, mid_num_let, single_quote])
        orelse ucd_is_category(CP, ['Mn', 'Me', 'Cf', 'Lm', 'Sk']).
```
where ```ucd_is_category/2```, ```ucd_has_property/2``` and ```ucd_word_break/2``` are generated by the parse transform.

It will be great if something similar to that parse transform is available in the OTP.

**Comment by dgud** [comment]
If you build this branch there is a 3rd low level file generated, unicode_util.erl.
The case information is available in that module.

`#{fold:="ss",lower:=223,title:="Ss",upper:="SS"} = unicode_util:get_case($ß)`

There is also a `#{ccc:=0, compat:=_, canon:=_} = unicode_util:lookup($å),`
which could include more information like category for each CodePoint.

But there is also size argument, currently unicode_util takes 925kB loaded into erts 
and we want to keep that as small as possible.

**Comment by jlouis** [comment]
Roughly stolen from the Standard ML basis are some functions you may want to consider for inclusion. They can probably be omitted in a first version, but don't construct your internals such that it makes it impossible to implement these I'd say:

* `str:fields/2`. Consider the case where you are splitting on `"|"` and you have the string `"a||b|c"`. Now should the output be `["a", "", "b", "c"]` or `["a","b","c"]` ? In the SML basis library, `fields` returns the former, whereas `tokens` returns the latter. The former is especially nice in the context of e.g., CSV-parsing, whereas the latter is good in typical tokenization where you want to strip empty matches.

* `str:append` `str:map/2` and `str:translate/3`. The map function is obvious. `translate` is the same as `append (map(F, Str))` where append is the obvious concatenation of strings. `translate` can often be made faster if explicit since you can avoid data buildup and fuse directly.

* `str:collate(F, S, T)` generalized compare where `F(CharS, CharT) -> bool()` is a `=<` on individual characters. Good when the normal sort order isn't what you want.

* `str:split` probably needs a [global] option like `binary:split/3` over time. Otherwise people end up implementing this themselves. OTOH, it is better the split a bit at a time for efficiency reasons, but... 

## Other considerations

* I think it is smart to create `str_lang` or `str_icu` as a separate module for mangling language specific stuff later on. This keeps the world in order: Computer Science in one module, Linguistics in another. Also, language stuff could be kept outside the basis OTP installation and added on an as-needed basis, since it will change over time and not necessarily on the OTP release cycle. Collation tables for general unicode comes to mind.

* In short: `str` should provide tooling for a `str_lang` module, but not the solution for its woes.

* Two data structures are often nice to have on strings: tries (spell-checking, compact representation) and mappings from strings to integers so you have a succinct representation of strings with repetition. They should not be part of this module, but thinking about support for these in other modules would be nice.

**Comment by isvilen** [comment]
That was my point - instead having one single big unicode_util.erl module providing some/all information from Unicode Character Database it is preferable to have a way to include only data relevant to specific use case.
If someone needs code points names there is no need that information to be loaded for users that do text normalization only. The approach using parse transformation provides this and also there is opportunities to optimize for size or performance based on the actual use case, for example code generated for ```ucd_is_category(CP, ['Mn', 'Me', 'Cf', 'Lm', 'Sk'])``` function needs only a small subset of information available in UnicodeData.txt file.

**Comment by dgud** [comment]
@jlouis 

str:split/3 works as you suggested:
```erlang
15> str:split("a||b|c", "|", all).      
["a",[],"b","c"]
```
I have not implemented searching for multiple strings in 'find' and 'split'.

I have avoided anything related to compare since that is often language dependent.
str:gc/1 and str:cp/1 gives you the possibility to build your own functions.

@isvilen Adding category and bidi makes unicode_util.beam 1196KB,
so I will not add that now and leave the parse_transform to you outside of the stdlib.

**Comment by asabil** [comment]
As @isvilen suggests, I think it would be much better to unicode_util into separate modules, util modules tend to become real kitchen sinks over time as we all know.

**Comment by okeuday** [comment]
@dgud If you could add a module version to the header of the unicode_util that provides the unicode version from the input files, that will provide an easy way to understand what is supported, which can help in the future, if/when it needs to be updated during runtime.  I think it is easier to keep it as a single module to keep it easy to update, but I can understand that people are concerned about the memory aspect of it.

**Comment by dgud** [comment]
@asabil There is currently no unnecessary data in unicode_util, there is case information for uppercase..casefold functions, grapheme cluster data for correctly combining codepoints to grapheme clusters and normalization data for unicode:characters_to_nf(k)[c|d] functions.
So the data is needed for the functionality exposed in the new functions.
It might be possible to do some memory vs speed optimizations/changes.

unicode_util was never intended to be an api, it just contains the data needed to for the new api functions in str and in unicode module. Maybe it is a bad name, maybe unicode_internal or unicode_data as it was called a week ago. The table lookup functions where there so I could debug it.

@okeuday there is a hardcoded unicode_util:spec_version() function.

**Comment by okeuday** [comment]
@dgud That is good, and that way stripping the beam won't get rid of the version.

**Comment by isvilen** [comment]
@dgud If unicode_util module is to be treated as internal module and not a public API then what is suggested way for an application to extend Unicode functionality provided by stdlib if needed?

**Comment by dgud** [comment]
@isvilen If there shall be no more data in stdlib, then is has to provided outside of the stdlib which fits with @jlouis suggestions. I don't really have a strong opinion.

File: lib/stdlib/src/str.erl [hunk 1]
```diff
@@ -0,0 +1,836 @@
...
+%% If these codepoints appear, functions like 'find/2' may return a string
+%% which starts inside a grapheme cluster.
+%% These exceptions are made because the codepoints classes are
+%% seldom used and require that we are able look at previous codepoints in
+%% the stream and is thus hard to implement effectively.
+%%
+%% GC (grapheme cluster) means that searching for [101,778] 'e̊' in the
+%% middle of a string of 'e's should find the 'e̊' not an 'e'
+%%
+%% find("eeeee̊eee", "e̊") -> "e̊ee".
```
Comment (avellable) [inline lib/stdlib/src/str.erl]:
```
find("eeeee̊eee", "e̊") -> "e̊eee".
```
?

Comment (dgud) [inline lib/stdlib/src/str.erl]: Maybe not the best example it sets up the next line, should probably use a tokens example there,
which works on a single character.
`str:tokens("eeeeee̊eee",   ["e̊"]) -> ["eeeee","eee"]`

**Comment by josevalim** [comment]
@dgud the unicode test files are duplicated in two directories. For example, GraphemeBreakTest is both in uc_spec and the test directory. Maybe you want to keep just one of them?

And for what is worth, we took the same decision in Elixir of keeping the unicode_util file focused. It can really grow in size if we attempt to embed the whole unicode database.

File: lib/stdlib/uc_spec/gen_unicode_mod.escript [hunk 2]
```diff
@@ -0,0 +1,834 @@
...
+    io:put_chars(Fd, "lookup(Codepoint) ->\n"
+                 "    {CCC,Can,Comp} = unicode_table(Codepoint),\n"
+                 "    #{ccc=>CCC, canon=>Can, compat=>Comp}.\n\n"),
+    io:put_chars(Fd, "-spec get_case(char()) -> #{'fold':=gc(), 'lower':=gc(), 'title':=gc(), 'upper':=gc()}.\n"),
+    io:put_chars(Fd, "get_case(Codepoint) ->\n"
+                 "    case case_table(Codepoint) of\n"
+                 "        {U,L} -> #{upper=>U,lower=>L,title=>U,fold=>L};\n"
+                 "        {U,L,T,F} -> #{upper=>U,lower=>L,title=>T,fold=>F}\n"
+                 "    end.\n\n"),
+    io:put_chars(Fd, "spec_version() -> {9,0}.\n\n\n"),
```
Comment (josevalim) [inline lib/stdlib/uc_spec/gen_unicode_mod.escript]: Maybe move this to a `-define` so it is more straight-forward to update when the unicode files are updated?

**Comment by josevalim** [comment]
@dgud have you considered other module names besides `str`? A "string" in Erlang is a list and the `str` module handles `unicode:chardata/2`. The shorter and unclear name can be prone to confusion. For example, someone may write a typespec saying that they expect a `string()` while they mean `unicode:chardata()`.

There are different solutions to the problems above. Options for the module name may be `ustring`, `utf8` or even `chardata`. It may also be worth adding a built-in type to typespecs called `chardata()`, making the `chardata()` type more accessible without forcing developers to reach into `unicode:chardata()`.

It may be worth discussing the name as I believe the question "why erlang has both `str` and `string`" may be asked a lot. :)

**Comment by dgud** [comment]
> @dgud the unicode test files are duplicated in two directories. For example, GraphemeBreakTest is both in uc_spec and the test directory. Maybe you want to keep just one of them?
> 

Yes, I'll need to fix that, I couldn't decide where I wanted them.

> And for what is worth, we took the same decision in Elixir of keeping the unicode_util file focused. It can really grow in size if we attempt to embed the whole unicode database.

I might have been inspired by the elixir code :-) 

> Maybe move this to a -define so it is more straight-forward to update it when the unicode files are updated? Otherwise I can see someone updating the files and forgetting to update the version hidden here. :)

I don't know if there is that much difference in updating a macro or a function definition..

> @dgud have you considered other module names besides str? A "string" in Erlang is a list and the str module handles unicode:chardata/2. The shorter and unclear name can be prone to confusion. For example, someone may write a typespec saying that they expect a string() while they mean unicode:chardata().
> 
> There are different solutions to the problems above. Options for the module name may be ustring, utf8 or even chardata. It may also be worth adding a built-in type to typespecs called chardata(), making the chardata() type more accessible without forcing developers to reach into unicode:chardata().
> 
> It may be worth discussing the name as I believe the question "why erlang has both str and string" may be asked a lot. :)

`str` was a working name for the current implementation, waiting for a better name but so far the suggestions have been few.

Of the above I like ustring best, utf8 is kind of wrong and chardata is not the first module a new user would start looking for "string" functions.

Your input triggered the suggestion that we maybe should add the functions to the current string module and deprecated the old functions.

More suggestions or opinions?

**Comment by josevalim** [comment]
@dgud just to clarify, you said:

> The new string module currently called str, operates on grapheme clusters, though it does not handle extended grapheme clusters, i.e. prepend class, nor classic or "nfd" Hangul.

This only applies to the `str` module. The NFC/NFD conversions on `unicode` will work on Hangul and others, correct?

The reason I am asking is because then we should be able to remove the unicode normalization stuff from Elixir (as well as the casing one) when we depend on Erlang 20. :)

**Comment by dgud** [comment]
@josevalim Yes correct (*), and length, reverse, slice and *case functions in str works correctly, the above applies to functions that are "searching" the input, for example tokens, find and split.

(*): unicode_util doesn't handle graphemes for all smiley classes correct, there are a couple tests that are currently ignored. But from what I could see in elixir's code you didn't bother about that either.

**Comment by josevalim** [comment]
> But from what I could see in elixir's code you didn't bother about that either.

That's because those classes were added on recent Unicode versions and I forgot to revisit our implementation. After your feedback this morning, I went ahead and implemented the remaining cases (as well as fixed some bugs). Now we pass GraphemeBreakTest.txt fully. Thanks for the heads up! :)

Unless you are talking about NormalizationTest.txt, I still have to verify those. Does this PR pass the NormalizationTest.txt suite fully then?

**Comment by dgud** [comment]
Yes, I don't pass the GraphemeBreakTest.txt, now I guess I'll have to fix that :-)

**Comment by josevalim** [comment]
If it helps, here is [our fix](https://github.com/elixir-lang/elixir/commit/3e394e8ef0acaed5bf583e301a5843333279e5ed). We pass an atom on the extended rules to know what we should trigger next. Here is [the suite](https://github.com/elixir-lang/elixir/commit/636c49fb5cf3a52a520233b2a0a9d7c1b651c829#diff-a23c6094d19d40addfbc06b7bc0e213aR12) (notice we skip rules with D800 which is surrogate).

**Comment by josevalim** [comment]
@dgud btw, what do you think of renaming `str:cp` and `str:gc` to `str:next_codepoint` and `str:next_grapheme` or something of sorts? When I see `cp`, the first thing that comes to mind is "copy" and for `gc` it is "garbage collect". I think the longer names would be beneficial for code readability. :)

**Comment by lexmag** [comment]
I'm wondering about `str` module name, maybe it should be named `strings` to be consistent with `lists`, `maps`?

**Comment by josevalim** [comment]
@lexmag there is a discussion about the module name a bit higher up. The plural and singular discussion is tricky because you will find examples for both: lists, maps, array, queue, binary, etc. Since we already have `string`, I tend to prefer to keep whatever we come up with as singular (my vote for now is `ustring`).

**Comment by lexmag** [comment]
@josevalim I see. Judging by recent addition of `maps` I assume there is preference to plural. 🤔
`ustring` still could be interpreted confusingly, my vote goes to `chardata`.

**Comment by dgud** [comment]
Added a couple of commits:

- Fixed emoji handling
- Removed files that where in two places.
- Renamed gc and cp as suggested above and updated the module description.

**Comment by whatyouhide** [comment]
I cast my not very meaningful vote for `chardata` as well if we can go with that, otherwise I think keeping it plural would make sense (as @lexmag mentioned, `maps` was recently added and used plural).

**Comment by crownedgrouse** [comment]
A translit/1 function would also be great to cast data to ASCII if possible,  like Linux iconv does (//TRANSLIT).

**Comment by brigadier** [comment]
Looks great!
Any ETA on merging it into master? Can't help waiting to test it.

**Comment by essen** [comment]
Great move on renaming str to string. :+1:

**Comment by dgud** [comment]
I have merged str into string module.

That required some renaming:
- new strip => trim
- new tokens => lexemes 
- new join is removed since lists:join did the same thing.

New functions:
- nth_lexeme
- detach  (name suggestion?)

Docs [http://erlang.org/~dgud/stdlib_html/string.html](http://erlang.org/~dgud/stdlib_html/string.html)

**Comment by josevalim** [comment]
@dgud those changes look fantastic. Some feedback:

  1. The first examples in the new module are missing some double quotes:

        <<"abc..åäö/utf8>>   is a valid string
        <<"abc..åäö>>        is NOT a valid string, latin1 encoded binary

  2. The docs say: "The functions may crash for invalid utf8-encoded binaries." I remember we talked about auto-synchronization. Are you planning to explicitly rely on it or are you planning to expect valid input mostly? I would believe such should be mentioned in the documentation.

  3. Alternative names for `string:detach/2` I could think of: "partition", as you are breaking it in two, or "take", as you are taking some characters, returning the characters taken and the ones that remain (similar to maps:take/2 in some sense).

  4. I would suggest for the examples or the documentation of "titlecase" to clarify what happens with the remaining of the string. For example, what is the output of "Hello World". Do you lowercase "W" or do you keep it as is?

  5. Some search functions use Haystack and Needle as the name of its arguments but I would say those names are rather implementation details and may be confusing to anyone who is not familiar with how searches work. I would suggest renaming them respectively to String and Pattern (or SearchPattern/LookupPattern/LookupString).

Finally, I would like to ask for the old functions in the string module to be deprecated only on OTP 22. The reason is that I believe developers working on libraries (or languages) would like to provide the same backwards compatibility period as OTP which is of 3 releases. If warnings are emitted only by Erlang 22, it means I can support Erlang 20, Erlang 21 and Erlang 22, all running without warnings.

However, if deprecation warnings are emitted already by Erlang 21, I will have to choose between no longer supporting 19, emitting warnings on 21, or adding conditional checks to my code.

**Comment by dgud** [comment]
@josevalim Thanks for the opinions and suggestions.

I have updated the documentation and renamed detach to take, partition takes a fun in lists and I want to leave the option to add the same functions to the string api.

In my opinion we must have the ability to deprecate stuff a year from now, otherwise it will take another year before people start to change their code.
You can suppress the deprecated warnings?

**Comment by josevalim** [comment]
> Thanks for the opinions and suggestions.

Glad to help! 😃 

> In my opinion we must have the ability to deprecate stuff a year from now, otherwise it will take another year before people start to change their code.

I would say this is exactly the reason for waiting more than one year. If you force everyone to upgrade already on OTP 21, when they upgrade, they will automatically restrict themselves to the last two OTP versions: 20 and 21. Instead I think we should push the ecosystem to support at least the last three versions.

Ultimately, folks will be able to find workarounds, but I also think we can lead with best practices.

**Comment by IngelaAndin** [comment]
Deprecating them now does not automatically mean we have to remove them in 21. I think we want the warnings now, and we remove them  some time later. The deprecating framework may be somewhat confusing but that how is what was done for old crypto interface for instance. Deprecated in 14 or 15 (cant remember) with warnings immediately and removed no in 20.

**Comment by josevalim** [comment]
Thanks @IngelaAndin, that's a very good point. While I assume most developers don't want their libraries and codebases running with warnings, if the warnings are emitted during compilation, I don't see any problem with deprecating it sooner.

**Comment by jlouis** [comment]
Historically, I think there are a lot of examples as to what Ingela is suggesting. You deprecate early on, and then you decide when to remove the feature later. For some things, the fix is rather straightforward. The replacement of `rand` for `random` is one example that is mostly mechanical and tend to be somewhat restricted in its use in code bases. You still get a random number, just better. In other cases, users might have relied a lot on a particular functionality, and this means the change could take more time to implement. `erlang:now()` is probably an example, where people have relied on its uniqueness property in their code.

In short, the more impact a change has the potential to have in a code base and how subtle it is to make the correct fix, the longer I think one should let a feature live in a deprecated state.

**Comment by bjorng** [comment]
`string:nth_lexeme/3` have some surprising return values:

```
1> string:nth_lexeme("str", 0, []).
[]
2> string:nth_lexeme("str", 1, []).
"str"
3> string:nth_lexeme("str", 2, []).
[]
4> string:nth_lexeme("str", 3, []).
[]
```

I would expect that the first expression (N==0) would return the input string.

Here are some more surprising return values:

```
5> string:nth_lexeme("str", 0, [$x]).
"str"
6> string:nth_lexeme("str", 1, [$x]).
"str"
7> string:nth_lexeme("str", 2, [$x]).
[]
8> string:nth_lexeme("str", 3, [$x]).
[]
```

I would have expected that the second expression (N==1) would return `[]`.

Also, the code clearly returns an empty list if you ask for token that does not exist, but that behavior is not documented.

**Comment by bjorng** [comment]
I would expect a crash:

```
1> string:equal(a, a).
true
```

**Comment by brigadier** [comment]
Might be it worth mentioning somewhere in documentation of this module that unicode strings with list represenation cannot be used in i/o and iolist_ functions. As it is not obvious and can result in unexpected problems later in translation time.

**Comment by dgud** [comment]
@brigadier ```io:put_chars(UnicodeList) or io:format(DEV,"~ts~n", [UnicodeList]) ``` works fine for unicode lists if the "device" accepts unicode characters. That have worked since ~R13, this PR doesn't change anything fundamental about io, erlang and unicode, it adds new functions that also handles utf8 encoded binaries as arguments and return values. Since many uses binaries to store strings instead of lists.

@bjorng equal does not look at the data if they are equal, the previous version didn't and
then if you do, how do you handle? 
```
Str = ["long lists of chars followed by something illegal"|atom],
string:equal(Str, Str)

```
Do you always verify everything in the string the other functions in this module doesn't?

**Comment by bjorng** [comment]
@dgud No, I don't mean that the other functions must verify all parts of its string argument.

But it was surprising to me that an equal function does not verify its arguments.

I now realize that `equal/2` would be a lot slower if it needed to verify the rest of the strings even if an early part of the strings were not equal. But perhaps it would be logical that `equal/2` would guarantee that the part of the strings that `equal/2` needs to look are indeed valid strings (i.e. if it returns `true`, both arguments are valid string). That would be similar to the other functions (they verify as much as of string that is needed).

**Comment by bjorng** [comment]
```
1> string:concat(<<233,165,155>>, []).
[<<233,165,155>>]
```

This is not a flat list or binary.

Also, I think it should be made clearer in the documentation that a flat result is only guaranteed if both arguments are flat to begin with.

**Comment by dgud** [comment]
@bjorng Thanks for looking, had forgot that concat was on TODO list to discuss.

**Comment by bjorng** [comment]
I have created a gist of my string_eqc.erl QuickCheck module:

https://gist.github.com/bjorng/03a869392d5a969ebf2c40044b664190

It found a quite a few bugs in `string`.
